hack/yaml-gen/util.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"encoding/json"
+	"sort"
+	"strconv"
+
+	"gopkg.in/yaml.v3"
+	v1ext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+var (
+	defaultTemplateFuncs = map[string]any{
+		"parseToConfigYaml":     parseToConfigYaml,
+		"parseToConfigYamlOpts": parseToConfigYamlOpts,
+		// this is an always false flag to give explicit intent
+		// to blocks such as {{ with deleteCurrent }}.
+		"deleteCurrent": func() bool { return false },
+	}
+)
+
+func parseToConfigYaml(data []byte) string {
+	parser := crdParser{}
+	yamlNode, err := parser.parseToYamlNode(data)
+	if err != nil {
+		panic(err)
+	}
+	return toYaml(yamlNode)
+}
+
+func parseToConfigYamlOpts(data []byte, noComments, noDefaults bool) string {
+	parser := crdParser{NoComments: noComments, NoDefaults: noDefaults}
+	yamlNode, err := parser.parseToYamlNode(data)
+	if err != nil {
+		panic(err)
+	}
+	return toYaml(yamlNode)
+}
+
+func toYaml(val interface{}) string {
+	result, err := yaml.Marshal(val)
+	if err != nil {
+		panic(err)
+	}
+	return string(result)
+}
+
+// These util functions are here to primarily deal with the default and example options,
+// they take the raw json and create a yaml node out of them.
+
+// Functions to help spit out string versions of examples to be used in comments
+func parseArrayJSONExample(rawJson *v1ext.JSON) string {
+	results := parseArrayJSONValue(rawJson)
+	if len(results) == 0 {
+		return ""
+	}
+	return toYaml(results)
+}
+func parseMapJSONExample(rawJson *v1ext.JSON) string {
+	results := parseMapJSONValue(rawJson)
+	if len(results) == 0 {
+		return ""
+	}
+	return toYaml(results)
+}
+func parseScalarJSONExample(rawJson *v1ext.JSON) string {
+	results := parseScalarJSONValue(rawJson)
+	if len(results) == 0 {
+		return ""
+	}
+	return toYaml(results)
+}
+
+// Best effort to parse raw json examples, controller-gen will break if the examples
+func parseArrayJSONValue(rawJson *v1ext.JSON) []*yaml.Node {
+	if rawJson == nil {
+		return nil
+	}
+	var arrayContainer []interface{}
+	if err := json.Unmarshal(rawJson.Raw, &arrayContainer); err != nil {
+		panic(err)
```

Comment:
Yeah that's what I was thinking, if there is bad input the controller-gen will fail to parse it before we do. In the case of mismatch, I updated the panic messages to include type information for easier debugging and a comment on maybe why we want to just panic in these functions. It does keep the code a little cleaner. If we aren't comfortable with it, we can bubble up the error with out a problem, the end result is the same since these functions will execute primarily in the template engine.

Reply:
Sounds like a plan, easy enough change if it comes to it, I did some quick testing and with the updated error messages, it should be clear where something went wrong.